python/pyspark/ml/torch/distributor.py

```diff
@@ -0,0 +1,491 @@
+        super().__init__(num_processes, local_mode, use_gpu, spark)
+        self.framework = framework
+        self.ssl_conf = "pytorch.spark.distributor.ignoreSsl"  # type: ignore
+        self._validate_input_params()
+
+    def _validate_input_params(self) -> None:
+        """Validates input params
+
+        Raises
+        ------
+        ValueError
+            Thrown when user fails to provide correct input params
+        """
+        super()._validate_input_params()
+        if self.framework not in self.available_frameworks:
+            raise ValueError(
+                f"{self.framework} is not a valid framework."
+                f"Available frameworks: {self.available_frameworks}"
+            )
+
+    def _run_local_training(
+        self,
+        framework_wrapper_fn: Optional[Callable],
+        train_fn: Union[Callable, str],
+        *args: Any,
+    ) -> Optional[Any]:
+        CUDA_VISIBLE_DEVICES = "CUDA_VISIBLE_DEVICES"
+        old_cuda_visible_devices = os.environ.get(CUDA_VISIBLE_DEVICES, "")
+        cuda_state_was_set = CUDA_VISIBLE_DEVICES in os.environ
+        if not framework_wrapper_fn:
+            raise RuntimeError("Unknown combination of parameters")
+        try:
+            if self.use_gpu:
+                if "gpu" not in self.sc.resources:
+                    raise ValueError('"gpu" was not found in the context resources.')
+                gpus_owned = get_gpus_owned(self.sc.resources["gpu"].addresses)
+
+                if self.num_processes > len(gpus_owned):
+                    raise ValueError(
+                        f"""{self.num_processes} processes were requested
+                        for local training with GPU training but only
+                        {len(gpus_owned)} GPUs were available."""
+                    )
+                selected_gpus = [str(e) for e in random.sample(gpus_owned, self.num_processes)]
+                os.environ[CUDA_VISIBLE_DEVICES] = ",".join(selected_gpus)
+
+            output = framework_wrapper_fn(self.__dict__, train_fn, *args)
+        finally:
+            if cuda_state_was_set:
+                os.environ[CUDA_VISIBLE_DEVICES] = old_cuda_visible_devices
+            else:
+                if (os.getenv(CUDA_VISIBLE_DEVICES, "")) != "":
+                    del os.environ[CUDA_VISIBLE_DEVICES]
+
+        return output
+
+    @staticmethod
+    def _run_training_on_pytorch_file(
+        input_params: dict[str, Any], train_path: str, *args: Any
+    ) -> None:
+        training_command = create_torchrun_command(input_params, train_path, *args)
+        execute_command(training_command)
+
+    def run(self, train_fn: Union[Callable, str], *args: Any) -> Optional[Any]:
+        """Runs distributed training on PySpark.
+
+        Parameters
+        ----------
+        train_fn : Union[Callable, str]
+            Either a PyTorch/PyTorch Lightning training function or the path to a python file
+            that launches distributed training.
```

Comment:
Param name is not good.
It looks like a python function but it can also be a path to a python code file.

Reply:
Would you suggest `train_object` or just `train`? I am also not entirely sure about what the run(...) API should look like. @lu-wang-dl any thoughts from your side?

Reply:
`train_object` is fine with me.